test/extensions/filters/http/ext_proc/filter_test.cc

```diff
@@ -2261,6 +2272,198 @@ TEST_F(HttpFilterTest, FailOnInvalidHeaderMutations) {
+          callbacks.addStreamDecoderFilter(mock_filter);
+        };
+        manager.applyFilterFactoryCb({}, mock_filter_cb);
+
+        return true;
+      }));
+  EXPECT_CALL(*mock_filter, decodeHeaders(_, false))
+      .WillOnce(Invoke([&](Http::RequestHeaderMap& headers, bool end_stream) {
+        // The next decoder filter should be able to see the mutations made by the external server.
+        EXPECT_FALSE(end_stream);
+        EXPECT_EQ(headers.Path()->value().getStringView(), "/mutated_path/bluh");
+        EXPECT_EQ(headers.get(Envoy::Http::LowerCaseString("foo"))[0]->value().getStringView(),
+                  "gift-from-external-server");
+        mock_filter->callbacks_->sendLocalReply(Http::Code::OK, "Direct response from mock filter.",
+                                                nullptr, absl::nullopt, "");
+        return Http::FilterHeadersStatus::StopIteration;
+      }));
+  EXPECT_CALL(response_encoder_, encodeHeaders(_, _))
+      .WillOnce(Invoke([&](const Http::ResponseHeaderMap& headers, bool end_stream) -> void {
+        EXPECT_FALSE(end_stream);
+        EXPECT_EQ(headers.Status()->value().getStringView(), "200");
+      }));
+  EXPECT_CALL(response_encoder_, encodeData(_, _))
+      .WillOnce(Invoke([&](Buffer::Instance& data, bool end_stream) -> void {
+        EXPECT_TRUE(end_stream);
+        EXPECT_EQ(data.toString(), "Direct response from mock filter.");
+      }));
+  // Start the request.
+  EXPECT_CALL(*codec_, dispatch(_)).WillOnce(Invoke([&](Buffer::Instance& data) -> Http::Status {
+    EXPECT_EQ(data.length(), 5);
+    data.drain(5);
+
+    HttpConnectionManagerImplMixin::decoder_ = &conn_manager_->newStream(response_encoder_);
+    Http::RequestHeaderMapPtr headers{new Http::TestRequestHeaderMapImpl{
+        {":authority", "host"}, {":path", "/bluh"}, {":method", "GET"}}};
+    HttpConnectionManagerImplMixin::decoder_->decodeHeaders(std::move(headers), false);
+    Buffer::OwnedImpl request_body("Definitely more than 10 bytes data.");
+    HttpConnectionManagerImplMixin::decoder_->decodeData(request_body, true);
+    // Now external server returns the request header response.
+    auto response = std::make_unique<ProcessingResponse>();
+    auto* headers_response = response->mutable_request_headers();
+    auto* hdr = headers_response->mutable_response()->mutable_header_mutation()->add_set_headers();
+    hdr->mutable_header()->set_key("foo");
+    hdr->mutable_header()->set_value("gift-from-external-server");
+    hdr = headers_response->mutable_response()->mutable_header_mutation()->add_set_headers();
+    hdr->mutable_header()->set_key(":path");
+    hdr->mutable_header()->set_value("/mutated_path/bluh");
+    HttpFilterTest::stream_callbacks_->onReceiveMessage(std::move(response));
+
+    return Http::okStatus();
+  }));
+
+  Buffer::OwnedImpl fake_input("hello");
+  conn_manager_->onData(fake_input, false);
```

Comment:
Suppose hypothetically that the bug was not fixed. In what way would you expect this test to fail? The test description mentions avoiding a 413 response, but it's not clear to me where that is verified in this test. I see the "200" response, but IIUC that is at the generation end, from the mock upstream filter.

Is it the case that the request wouldn't even be seen by `mock_filter_` if the bug were present, so the fact that `mock_filter_` sees a call shows that we've avoided the high watermark call as intended? If so, I think that's definitely worth a comment in the test, since it's not obvious.

Reply:
if it fails, there isn't gonna be a "gift-from-external-server", the previous highwatermark breach would end up with a 500 response too large error directly in onData call, while waiting for the responseHeaders response from ext-server.

Reply:
updated the comment at line 2387.

Reply:
LG